pkg/networkpath/traceroute/common/common.go

```diff
@@ -71,90 +92,217 @@ func (c CanceledError) Error() string {
+	p.icmpResponse.TypeCode = p.icmpLayer.TypeCode
 
 	var icmpPayload []byte
-	if len(icmpv4Layer.Payload) < 40 {
-		log.Tracef("Payload length %d is less than 40, extending...\n", len(icmpv4Layer.Payload))
+	if len(p.icmpLayer.Payload) < 40 {
+		log.Tracef("Payload length %d is less than 40, extending...\n", len(p.icmpLayer.Payload))
 		icmpPayload = make([]byte, 40)
-		copy(icmpPayload, icmpv4Layer.Payload)
-		// we have to set this in order for the TCP
+		copy(icmpPayload, p.icmpLayer.Payload)
+		// we have to set this in order for the inner
 		// parser to work
 		icmpPayload[32] = 5 << 4 // set data offset
 	} else {
-		icmpPayload = icmpv4Layer.Payload
+		icmpPayload = p.icmpLayer.Payload
 	}
 
 	// a separate parser is needed to decode the inner IP and TCP headers because
 	// gopacket doesn't support this type of nesting in a single decoder
-	var innerIPLayer layers.IPv4
-	var innerTCPLayer layers.TCP
-	innerIPParser := gopacket.NewDecodingLayerParser(layers.LayerTypeIPv4, &innerIPLayer, &innerTCPLayer)
-	if err := innerIPParser.DecodeLayers(icmpPayload, &decoded); err != nil {
+	innerIPParser := gopacket.NewDecodingLayerParser(layers.LayerTypeIPv4, &p.innerIPLayer, &p.innerTCPLayer)
+	if err := innerIPParser.DecodeLayers(icmpPayload, &p.decoded); err != nil {
 		return nil, fmt.Errorf("failed to decode inner ICMP payload: %w", err)
 	}
-	icmpResponse.InnerSrcIP = innerIPLayer.SrcIP
-	icmpResponse.InnerDstIP = innerIPLayer.DstIP
-	icmpResponse.InnerSrcPort = uint16(innerTCPLayer.SrcPort)
-	icmpResponse.InnerDstPort = uint16(innerTCPLayer.DstPort)
-	icmpResponse.InnerSeqNum = innerTCPLayer.Seq
+	p.icmpResponse.InnerSrcIP = p.innerIPLayer.SrcIP
+	p.icmpResponse.InnerDstIP = p.innerIPLayer.DstIP
+	p.icmpResponse.InnerSrcPort = uint16(p.innerTCPLayer.SrcPort)
+	p.icmpResponse.InnerDstPort = uint16(p.innerTCPLayer.DstPort)
+	p.icmpResponse.InnerIdentifier = p.innerTCPLayer.Seq
 
-	return &icmpResponse, nil
+	return p.icmpResponse, nil
 }
 
+func (p *ICMPParser) parseWithInnerUDP(payload []byte) (*ICMPResponse, error) {
+	p.decoded = []gopacket.LayerType{}
+	if err := p.packetParser.DecodeLayers(payload, &p.decoded); err != nil {
+		return nil, fmt.Errorf("failed to decode ICMP packet: %w", err)
+	}
+	// since we ignore unsupported layers, we need to check if we actually decoded
+	// anything
+	if len(p.decoded) < 1 {
+		return nil, fmt.Errorf("failed to decode ICMP packet, no layers decoded")
+	}
+	p.icmpResponse.TypeCode = p.icmpLayer.TypeCode
+
+	p.icmpResponse.InnerSrcIP = p.innerIPLayer.SrcIP
+	p.icmpResponse.InnerDstIP = p.innerIPLayer.DstIP
+	p.icmpResponse.InnerSrcPort = uint16(p.innerUDPLayer.SrcPort)
+	p.icmpResponse.InnerDstPort = uint16(p.innerUDPLayer.DstPort)
+	// the packet's checksum is used as the identifier for UDP packets
+	p.icmpResponse.InnerIdentifier = uint32(p.innerUDPLayer.Checksum)
+
+	return p.icmpResponse, nil
+}
+
+// ParseICMP takes in an IPv4 header and payload and tries to convert to an ICMP
```

Comment:
I'm going to be deleting this whole block. The functionality has been refactored to the ICMPParser struct for what existed, and the inner UDP parsing code has been added